### ISSUE TITLE
Multigrid  transfer operators for serendipity elements

### ISSUE THREAD

**Issue by rckirby** [firedrakeproject/firedrake#1978]
What is required to implement "native" prolongation/restriction operators serendipity elements rather than going through embedding into the Q space?  I have looked at the code a bit, but am finding it difficult to get started.

**Comment by wence-** [comment]
First thing. Add the canonical name of the serendipity elements to the set of `native` elements in mg/embedded.py. That might be enough. Assuming all the nodes are point evaluation, the rest of it might work correctly.

**Comment by rckirby** [comment]
The nodes are linear combinations of point evaluations (some quadrature) but a few (vertices) are evaluation at a single point.  Changing to `native = frozenset(["Lagrange", "Discontinuous Lagrange", "Real", "Q", "DQ", "S"])` gives horrible numerical behavior on one mesh and then a failure of the form: 
```Could not identify cell in transfer operator. Point: 0.00000000000000e+00 0.00000000000000e+00 
Number of candidates: 1. Best distance located:   1.200000e+01
```
on the next mesh.

**Comment by wence-** [comment]
> The nodes are linear combinations of point evaluations (some quadrature) but a few (vertices) are evaluation at a single point. 

OK, this will just not work with the current setup. Presently the code for grid transfer (other than injection of DG fields) only handles point evaluation nodes, because the way it works is (to take the example of prolongation) to:
1. Walk over the fine dofs one-by-one
2. At each dof, compute the value by hitting the coarse function with the fine dual. But the code assumes that this dual is point evaluation (so it just evaluates the coarse function at the physical location of the fine grid dual node, which is obviously nonsense for anything other than point evaluations)

**Comment by rckirby** [comment]
One *could* rewrite serendipity elements to use point evaluations.  A bunch of work at this point.  But after proper dual evaluation lands, would it be possible to handle this then?

**Comment by wence-** [comment]
> One could rewrite serendipity elements to use point evaluations. 

I should have been more restrictive. The code presently only works for those functionals that are a _single_ point evaluation.

> A bunch of work at this point. But after proper dual evaluation lands, would it be possible to handle this then?

Possible, sure. It's not straightforward, because there are more moving parts: the updates to dual evaluation and interpolation are going to make it cleaner to do that stuff _on a single mesh_. But we have two meshes involved here. So matching things up is not as easy.

**Comment by rckirby** [comment]
Yes, one *can* write a serendipity basis dual to point evaluations.  It's just not the one that we have.

I think I see some of the issues -- you do need some kind of pullback business between coarse and fine to properly apply the dofs.  Probably that can be made to work in some generality.  But it will take some work.
